### ISSUE TITLE
Non-Symmetric Matrix Structure in Two-Phase Simulation

### ISSUE THREAD

**Issue by bska** [OPM/opm-simulators#665]
I tried to test @babrodtk's changes OPM/opm-core#1003 and #661 on a two-phase simulation case that is currently **not** in [opm-data](https://github.com/OPM/opm-data).  I can provide additional details offline.  Towards the end of the simulation I get an exception from Dune ISTL's [`compressed_base_array_unmanaged<>`](https://gitlab.dune-project.org/core/dune-istl/blob/master/dune/istl/basearray.hh#L565-571) saying

```
index 2345 not in compressed array
```

This is just the symptom, of course.

The immediate calling code is method

``` C++
NewtonIterationBlackoilInterleavedImpl<>::formInterleavedSystem()
```

and especially the [loop](https://github.com/OPM/opm-simulators/blob/master/opm/autodiff/NewtonIterationBlackoilInterleaved.cpp#L349-L364) that inserts derivatives into the interleaved system matrix.

Just prior to encountering this problem all wells were switched from being constrained by `RESERVOIR_RATE` targets to being constrained by `BHP` targets

```
Iter  W-FLUX(W)  W-FLUX(O)
   0  1.050e-05  6.617e-04
well converged iter: 0

Iter   MB(Wat)    MB(Oil)     CNV(W)     CNV(O)   W-FLUX(W)  W-FLUX(O)
   0  6.382e-06  3.804e-07  3.212e-03  2.453e-03  1.050e-05  6.617e-04
Switching control mode for well BR-P-1 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-2 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-3 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-4 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-5 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-6 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-7 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-8 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-9 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-10 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-11 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-12 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-13 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-14 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-16 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-17 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-18 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-19 from RESERVOIR_RATE to BHP
Switching control mode for well BR-P-20 from RESERVOIR_RATE to BHP
   1  1.219e-07  5.592e-06  2.465e-03  7.161e-03  1.401e-04  1.158e-02
```

I have pasted a stack trace of the frames that led to this point below, but I'm not really sure how to proceed.  Getting to this point, which is also not entirely repeatable, in debug mode requires about 18 hours on my computer so I'll leave it in the debugger if anyone have any immediate suggestions.  Otherwise, I suggest we make a note of this fact and engage in further analysis once the restart feature is more fully baked so that we can start debugging from the last converged solution.

---

```
(gdb) bt
#0  Dune::compressed_base_array_unmanaged<Dune::MatrixBlock<float, 2, 2>, std::allocator<Dune::MatrixBlock<float, 2, 2> > >::operator[] (this=0xf6a0cf0, i=2345) at /home/bska/work/opm/inst/dune/2.3.1/dbg/seq/include/dune/istl/basearray.hh:571
#1  0x0000000000fb14b8 in Opm::NewtonIterationBlackoilInterleavedImpl<2, float>::formInterleavedSystem (this=0x51c27b0, eqs=std::__debug::vector of length 2, capacity 2 = {...}, istlA=...) at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/NewtonIterationBlackoilInterleaved.cpp:360
#2  0x0000000000fa8675 in Opm::NewtonIterationBlackoilInterleavedImpl<2, float>::computeNewtonIncrement (this=0x51c27b0, residual=...) at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/NewtonIterationBlackoilInterleaved.cpp:428
#3  0x0000000000fa15ec in Opm::NewtonIterationBlackoilInterleaved::computeNewtonIncrement (this=0x37ce510, residual=...) at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/NewtonIterationBlackoilInterleaved.cpp:572
#4  0x0000000000eaa053 in Opm::BlackoilModelBase<UnstructuredGrid, Opm::StandardWells, Opm::BlackoilModel<UnstructuredGrid> >::solveJacobianSystem (this=0x9455e60) at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/BlackoilModelBase_impl.hpp:1418
#5  0x0000000000e8d911 in Opm::BlackoilModelBase<UnstructuredGrid, Opm::StandardWells, Opm::BlackoilModel<UnstructuredGrid> >::nonlinearIteration<Opm::NonlinearSolver<Opm::BlackoilModel<UnstructuredGrid> > > (this=0x9455e60, iteration=1, dt=75356.821814893832, nonlinear_solver=..., reservoir_state=..., 
    well_state=...) at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/BlackoilModelBase_impl.hpp:275
#6  0x0000000000e718e8 in Opm::NonlinearSolver<Opm::BlackoilModel<UnstructuredGrid> >::step (this=0x3777590, dt=75356.821814893832, reservoir_state=..., well_state=...) at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/NonlinearSolver_impl.hpp:99
#7  0x0000000000e8cdb2 in Opm::AdaptiveTimeStepping::stepImpl<Opm::NonlinearSolver<Opm::BlackoilModel<UnstructuredGrid> >, Opm::BlackoilState, Opm::WellStateFullyImplicitBlackoil> (this=0x1a9e590, simulatorTimer=..., solver=..., state=..., well_state=..., outputWriter=0x3767d50)
    at /data/projects/opm/build/opm/dbg/seq/dune-2.3.1/opm-common/../inst/include/opm/core/simulator/AdaptiveTimeStepping_impl.hpp:166
#8  0x0000000000e717c8 in Opm::AdaptiveTimeStepping::step<Opm::NonlinearSolver<Opm::BlackoilModel<UnstructuredGrid> >, Opm::BlackoilState, Opm::WellStateFullyImplicitBlackoil> (this=0x1a9e590, simulatorTimer=..., solver=..., state=..., well_state=..., outputWriter=...)
    at /data/projects/opm/build/opm/dbg/seq/dune-2.3.1/opm-common/../inst/include/opm/core/simulator/AdaptiveTimeStepping_impl.hpp:117
#9  0x0000000000e52039 in Opm::SimulatorBase<Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >::run (this=0x3768010, timer=..., state=...) at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/SimulatorBase_impl.hpp:176
#10 0x0000000000e3dda2 in Opm::FlowMainBase<Opm::FlowMain<UnstructuredGrid, Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >, UnstructuredGrid, Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >::runSimulator (this=0x7fffffffdd10)
    at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/FlowMain.hpp:665
#11 0x0000000000e2c50c in Opm::FlowMainBase<Opm::FlowMain<UnstructuredGrid, Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >, UnstructuredGrid, Opm::SimulatorFullyImplicitBlackoil<UnstructuredGrid> >::execute (this=0x7fffffffdd10, argc=2, argv=0x7fffffffdfa8)
    at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/FlowMain.hpp:165
#12 0x0000000000e0bb7f in main (argc=2, argv=0x7fffffffdfa8) at /home/bska/work/opm/src/opm/opm-simulators/examples/flow.cpp:41
(gdb) f 1
#1  0x0000000000fb14b8 in Opm::NewtonIterationBlackoilInterleavedImpl<2, float>::formInterleavedSystem (this=0x51c27b0, eqs=std::__debug::vector of length 2, capacity 2 = {...}, istlA=...) at /home/bska/work/opm/src/opm/opm-simulators/opm/autodiff/NewtonIterationBlackoilInterleaved.cpp:360
360                             istlA[row][col][p1][p2] = sa[elem_ix];
```

**Comment by totto82** [comment]
I experience a similar symptom when I run the solvent model for Model 2. I got rid of the symptoms by adding all the derivatives in the "loop". I have not fully investigated what's causing the extra derivatives for the solvent model, but I suspect the wells.  

Restart should work, at least for three-phase problems.

**Comment by andlaus** [comment]
this seems to be fixed with #661?
